File: leetcode/Rotate List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (NULL == head)
        {
    		return head;
		}
        int length = 1;
		ListNode *end = head;
		while (end->next)
		{
			length ++;
			end = end->next;
		}
		if (0 == length)
		{
			return head;
		}
		k = k % length;
		if (0 == k)
		{
			return head;
		}
		ListNode *new_end = head;
		for (int i = 0; i < length - k - 1; ++ i)
		{
			new_end = new_end->next;
		}
		ListNode *new_head = new_end->next;
		end->next = head;
		new_end->next = NULL;
		return new_head;
    }
};

File: answers/leetcode/Rotate List/Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return null;
        }

        int length = 1;
        ListNode rear = head;
        while (rear.next != null) {
            length++;
            rear = rear.next;
        }
        k %= length;

        ListNode right = head;
        for (int i = 0; i < k; i++) {
            right = right.next;
        }

        ListNode left = head;
        while (right.next != null) {
            left = left.next;
            right = right.next;
        }

        rear.next = head;
        ListNode newHead = left.next;
        left.next = null;

        return newHead;
    }
}
//Runtime: 0 ms, faster than 100.00% of Java online submissions for Rotate List.
//Memory Usage: 38.1 MB, less than 96.91% of Java online submissions for Rotate List.
